### PR TITLE
Blocks are complete in program if they have no questions

### LINKED ISSUE #1877: Unable to submit completed application with ROC test program
So far only reproducible with a specific ROC program, config state to be shared to help with reproing in dev.

Repro steps with test program:
- fill out program questionnaire
- review screen has 'continue', clicking it takes the user to the same review screen

expected:
review screen has 'submit' button

### PR DESCRIPTION
**Pull request by bion**
To determine if an application is ok to submit or requires more answers from the applicant, we count the number of active blocks (i.e. blocks that are not hidden by predicate logic) and compare that with the number of active blocks that are completed (answered) or skipped if Optional. This caused a bug where blocks that have no questions would count towards the total of active but incomplete blocks, making it impossible to complete the application.

This change alters the logic for calculating block completion state to be true if the block has no questions.

Verified the updated test catches this bug:

image

Fixes #1877

### CODE REVIEW

File: universal-application-tool-0.0.1/app/services/applicant/Block.java [hunk 1]
```diff
@@ -205,8 +205,9 @@ public boolean isCompletedInProgramWithoutErrors() {
   }
 
   private boolean isCompleteInProgram() {
-    return getQuestions().stream()
-        .anyMatch(question -> question.isAnsweredOrSkippedOptionalInProgram());
+    return getQuestions().size() == 0
```
Comment (shanemc-goog) [inline universal-application-tool-0.0.1/app/services/applicant/Block.java:208]: GetQuestions() -> ImmutableLIst which has isEmpty().

Maybe add a javadoc that allows for no questions.

File: browser-test/src/application_review.test.ts [hunk 2]
```diff
@@ -85,6 +86,10 @@ describe('normal application flow', () => {
       'monthly-income-q',
     ])
 
+    // Intentionally add an empty block to ensure that empty blocks do not
```
[the remark this answers is not in the document]
Comment (bion) [reply browser-test/src/application_review.test.ts:89]: > This existing test is gross.

why?

> I also feel like the use case is an edge case (EG not a conscious part of the design) compared to the rest of this test, that warrants its own test also.

> FWIW, I don't think ind test framework setup/teardown affects the test run timing as much as the tests interacting with the UI.

I don't agree, we saw a pretty noticeable improvement in test runtime when we merged a bunch of smaller browser tests into single longer ones. I'd prefer to leave it in that condition for now.

The best way to improve the situation overall IMO would be to split out multiple CI tasks for running different batches of the browser tests in parallel. With that in place it'd probably be a lot less painful to have a bunch of smaller, more focused browser tests.

File: universal-application-tool-0.0.1/app/services/applicant/Block.java [hunk 3]
```diff
@@ -205,8 +205,9 @@ public boolean isCompletedInProgramWithoutErrors() {
...
+        || getQuestions().stream()
+            .allMatch(question -> question.isAnsweredOrSkippedOptionalInProgram());
```
[the remark this answers is not in the document]
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/Block.java:216]: Yes, I believe so, but I've not had time to investigate.

I haven't had time to investigate it more deeply though. Intuitively, a block should only be considered complete if all of its required questions are answered or intentionally skipped.

The browser tests passing with this change give me confidence to ship it.

File: universal-application-tool-0.0.1/app/services/applicant/Block.java (hunk 1 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/Block.java]: done

File: browser-test/src/application_review.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (shanemc-goog) [reply browser-test/src/application_review.test.ts:89]: Gross:  It's doing a lot a lot of things which makes it unfocused as far as test coverage and confidence go as well readability, which makes updating it error prone.

What is "normal" and how would a new dev know if they should update "normal"
 
This test is also both verifying validation concerns and resolution concerns some questions, but it's hard to know what the coverage is from the method naming/results list. Those concerns should ideally be more unit ized so that one question failure doesn't affect all question verification status.

As for the blocks, it's unclear why the blocks are organized that way, such that the testing they are doing is easily lost through future changes.  Aside from "More than 1 block" I don't know why there's 5 blocks versus 2.


Test Timing:

Is it possible that the time was improved because of sharing the Admin/Program setup, and not the user interaction with the program itself?

EG: we can have separate tests, share the same Program setup.  I've seen where it's all or nothing and not the middle ground.

[the remark this answers is not in the document]
Comment (shanemc-goog) [reply browser-test/src/application_review.test.ts:89]: Here's an example of sharing setup I did with the currency question type

https://github.com/seattle-uat/civiform/blob/6724de093b884d32af4ded5c38404fc84c25501d/browser-test/src/currency.test.ts#L47

File: universal-application-tool-0.0.1/app/services/applicant/Block.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (shanemc-goog) [reply universal-application-tool-0.0.1/app/services/applicant/Block.java:216]: Yours def. looks right.  I wonder if this is (part of) the issue Claudia surfaced in the email last week.

**Review by shanemc-goog — APPROVED** [review]

File: browser-test/src/application_review.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bion) [reply browser-test/src/application_review.test.ts:89]: captured this in a tech debt item to address down the line
